src/sdk/python/tests/iotea/core/util/test_mqtt_client.py

```diff
@@ -0,0 +1,342 @@
+
+
+@pytest.mark.parametrize('topic_ns', ['namespace/'])
+@pytest.mark.parametrize('topic', ['test_topic'])
+@pytest.mark.parametrize('message', ['{"test": "message"}'])
+@pytest.mark.asyncio
+#send a publish message, disconnect and reconnect, send another publish message
+#check if the client functions after disconnect and reconnect
+# pylint: disable=redefined-outer-name
+async def test_reconnect(test_case, mocker, topic_ns, topic, message):
+    mocker.patch('hbmqtt.client.MQTTClient.connect', wraps=mock_connect)
+    mocker.patch('hbmqtt.client.MQTTClient.publish', wraps=mock_publish)
+    mocker.patch('hbmqtt.client.MQTTClient.reconnect', wraps=mock_reconnect)
+    mocker.patch('hbmqtt.client.MQTTClient.disconnect', wraps=mock_disconnect)
+
+    mqtt_client = MqttClient('mqtt://localhost:1883', topic_ns, check_mqtt5_compatibility=False)
+
+    await mqtt_client.publish(topic, message, options=None)
+
+    assert mqtt_client.client.connect.call_count == 1
+    assert mqtt_client.client.publish.call_count == 1
+    mqtt_client.client.publish.assert_called_once_with(f'{topic_ns}{topic}', message.encode('utf-8'), qos=QOS_0, retain=False)
+
+    await mqtt_client.disconnect()
+    assert mqtt_client.client.disconnect.call_count == 1
+
+    await mqtt_client.client.reconnect(cleansession=True)
+
+    assert mqtt_client.client.disconnected is False
+
+    message = 'new message'
+    await mqtt_client.publish(topic, message, options=None)
+
+    assert mqtt_client.client.connect.call_count == 1
+    assert mqtt_client.client.publish.call_count == 2
+    mqtt_client.client.publish.assert_called_with(f'{topic_ns}{topic}', message.encode('utf-8'), qos=QOS_0, retain=False)
+
+# pylint: disable=redefined-outer-name
+def test_wrong_namespace(test_case):
+    topic_ns = 'no_trailing_slash'
+    with pytest.raises(Exception) as exc_info:
+        MqttClient('mqtt://localhost:1883', topic_ns, check_mqtt5_compatibility=False)
+    assert f'Given topic namespace {topic_ns} is invalid. It has to have a trailing slash' == str(exc_info.value)
+
+
+@pytest.mark.parametrize('topic_ns', ['namespace/'])
+@pytest.mark.parametrize('topic', ['test_topic'])
+@pytest.mark.parametrize('message', ['invalid_json_message'])
+@pytest.mark.asyncio
+async def test_validate_json(test_case, topic_ns, topic, message):
+    mqtt_client = MqttClient('mqtt://localhost:1883', topic_ns, check_mqtt5_compatibility=False)
+    with pytest.raises(Exception) as exc_info:
+        await mqtt_client.publish_json(topic, message)
+    assert f'Given JSON document is neither a dictionary nor a list' == str(exc_info.value)
+
+
+@pytest.mark.parametrize('topic_ns', ['namespace/'])
+@pytest.mark.parametrize('topic', ['test_topic'])
+@pytest.mark.parametrize('message', ['{"test": "message"}'])
+@pytest.mark.asyncio
+async def test_multiple_connects(test_case, mocker, topic_ns, topic, message):
+    connect_timeout = 2
+    async def mock_connect_block(broker_url, cleansession=True):
```

Comment:
Instead of relying on `asyncio.sleep()` consider trying `asyncio.Event` to synchronize and gain more control. I haven't tried, but maybe something like this is possible.

```python
async def mock_connect_block(alpha, beta):
    async def f(broker_url, cleansession=True):
        alpha.wait()
        beta.set()
    return f

alpha = asyncio.Event()
beta = asyncio.Event()

mocker.patch('hbmqtt.client.MQTTClient.connect', wraps=mock_connect_block(alpha, beta))

# test what you want to test while the client is busy connecting

alpha.set() #  unblocks the mocked connect method
beta.wait() #  wait for the mocked method to signal that it's done

# test what you want to test after the client is done connecting
```

Reply:
Maybe this ^^^ is fragile too. It depends on whether the test code and the connect function are running in the same thread. If they are, then the `beta` event can be removed because coroutines run to completion or `wait` before any other coroutine in the same thread can run. If they are not, then there's a small window between the calls to `beta.set()` and `beta.wait()` in which a context switch would potentially allow the test to continue to run before the MQTT client has had time to register that the mocked connect method has returned.

Reply:
Refactored.